### PR TITLE
Override MemoryStream.CopyTo and expose the override for netcoreapp1.1

### LINKED ISSUE #11723: CopyTo should be overriden in MemoryStream
I just made a change in CoreCLR (dotnet/coreclr#7131) that overrides `MemoryStream.CopyTo` since that method of Stream is now virtual. However it looks like since the copy of `MemoryStream` in here (which is used in corert) is separated from its parent which lives in [corert itself](https://github.com/dotnet/corert/blob/master/src/System.Private.CoreLib/src/System/IO/Stream.cs), the compiler will not be able to override the method here since it is not yet virtual in the contracts. When it is, the changes from that PR should be ported here.

@stephentoub @jkotas

### PR DESCRIPTION
**Pull request by jamesqo**
This change overrides `MemoryStream.CopyTo`, porting the changes from dotnet/coreclr#7131 and dotnet/coreclr#7579, exposes the override for netcoreapp1.1, and adds tests for the change.

Note that the build will fail until, as @jkotas said in #11723 the version of corert System.Private.CoreLib built against is updated. So please don't merge this yet.

Fixes #11723

@jkotas, @stephentoub, @weshaggard, @hughbe, @justinvp

**edit:** Aha, I just realized that the `Stream` ref class was actually hiding in [System.Runtime.cs](https://github.com/dotnet/corefx/blob/master/src/System.Runtime/ref/System.Runtime.cs) which is why I couldn't find it in System.IO.cs. Going to update shortly.

### CODE REVIEW

File: src/System.IO/tests/MemoryStream/MemoryStreamTests.cs [hunk 1]
```diff
@@ -194,5 +195,93 @@ public static void MemoryStream_WriteToTests_Negative()
...
+            var data3 = data2;
+            var stream3 = new MemoryStream(data3) { Position = data3.Length + 1 };
+
+            yield return new object[] { stream3, Array.Empty<byte>() };
+        }
+    }
+
+    public static class StreamExtensions
+    {
+        public static byte[] ToByteArray(this Stream stream)
```
Comment (stephentoub) [inline src/System.IO/tests/MemoryStream/MemoryStreamTests.cs]: Why is this necessary?  I only see it being used with MemoryStream, which already has ToArray.

Comment (jamesqo) [inline src/System.IO/tests/MemoryStream/MemoryStreamTests.cs]: Ah, I somehow forgot about that api. Thanks for reminding me.

File: src/System.IO/src/System/IO/MemoryStream.cs [hunk 2]
```diff
@@ -474,41 +474,46 @@ public override int ReadByte()
             return _buffer[_position++];
         }
 
-
-        public override Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken)
+        public override void CopyTo(Stream destination, int bufferSize)
```
Comment (weshaggard) [inline src/System.IO/src/System/IO/MemoryStream.cs]: In order to fix your build issues you are going to need a netcoreapp1.1 build configuration for the System.IO src project.

**Comment by weshaggard** [comment]
@ericstj do we have any other netcoreapp1.1 src builds that @jamesqo could use for reference?

**Comment by jkotas** [comment]
@jamesqo You may want to do this in two steps:

1. Make PR with just making the member virtual in System.Runtime.cs and get it merged. 
2. Wait for a few days to let the change to propagate through the system (for one of those version update
3. Do the rest of the changes.

(There is way to do it in one step, but it requires doing temporary changes that have to be removed in cleanup PR later once the change propagate through the system. So you can't really get away from doing this in two steps anyway... .)

**Comment by jamesqo** [comment]
@jkotas, OK, have opened #12876

**Comment by jamesqo** [comment]
@dotnet-bot test this please

**Comment by jamesqo** [comment]
@dotnet-bot test this please

**Comment by jamesqo** [comment]
@weshaggard @jkotas, it appears that the builds are still failing because `CopyTo` is not marked virtual. Any ideas?

File: src/System.IO/src/System.IO.builds [hunk 3]
```diff
@@ -10,6 +10,9 @@
       <TargetGroup>net463</TargetGroup>
     </Project>
     <Project Include="System.IO.csproj">
+      <TargetGroup>netcoreapp1.1</TargetGroup>
```
Comment (jkotas) [inline src/System.IO/src/System.IO.builds]: This should not be necessary - the default should be treated as `netcoreapp1.1`

File: src/System.IO/ref/System.IO.cs [hunk 4]
```diff
@@ -186,6 +186,9 @@ public partial class MemoryStream : System.IO.Stream
         public override long Position { get { throw null; } set { } }
         public override System.IAsyncResult BeginRead(byte[] buffer, int offset, int count, System.AsyncCallback callback, object state) { throw null; }
         public override System.IAsyncResult BeginWrite(byte[] buffer, int offset, int count, System.AsyncCallback callback, object state) { throw null; }
+#if netcoreapp11
```
Comment (jkotas) [inline src/System.IO/ref/System.IO.cs]: In `src\System.IO\ref\System.IO.csproj`, change:

`<NuGetTargetMoniker>.NETStandard,Version=v1.7</NuGetTargetMoniker>`

to 

`<NuGetTargetMoniker Condition="'$(TargetGroup)' == ''">.NETStandard,Version=v1.7</NuGetTargetMoniker>`.

It is necessary to pickup your earlier change.

Comment (jamesqo) [inline src/System.IO/ref/System.IO.cs]: @jkotas Thanks, your suggestions seem to have done the trick.

File: src/System.IO/src/System.IO.builds (hunk 3 above)
Comment (weshaggard) [inline src/System.IO/src/System.IO.builds]: @jkotas I don't think that is actually true. The default is netstandard1.7 (see https://github.com/dotnet/corefx/blob/master/src/System.IO/src/System.IO.csproj#L10).  There will need to be a netcoreapp1.1 configuration.

File: src/System.IO/src/System/IO/MemoryStream.cs [hunk 5]
```diff
@@ -474,41 +474,48 @@ public override int ReadByte()
             return _buffer[_position++];
         }
 
-
-        public override Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken)
+#if netcoreapp11
```
Comment (weshaggard) [inline src/System.IO/src/System/IO/MemoryStream.cs]: I would expect this #ifdef to be needed. I'm not sure how this PR even compiles because this method is only virtual and thus can only be override in the netcoreapp1.1 version of System.Runtime which would need to be referenced here. However as far as I can tell it is only referencing the netstandard1.7 version of the System.Runtime which isn't virtual. I'm going to see if I can build this locally to see why it didn't fail.

Comment (jkotas) [inline src/System.IO/src/System/IO/MemoryStream.cs]: `System.IO\src\System.IO.csproj` is referencing CoreLib ... it is why it compiles, and why I have said that netcoreapp1.1 is the "default".

Comment (weshaggard) [inline src/System.IO/src/System/IO/MemoryStream.cs]: Your correct this is working because it is directly compiling against System.Private.CoreLib. However I wouldn't say that makes it default to netcoreapp1.1 because it is still being packaged as netstandard1.7. For now I think we can live with this because we are soon going to overhaul our packages and this will become a non-issue. 

FYI @ericstj
